#pragma once

#include <stdint.h>

#define BAUDRATE 115200

#pragma pack(push, 1)
typedef struct _ValueInput {
  uint16_t estop_l;
  uint16_t estop_r;
  uint16_t sw_green;
  uint16_t sw_red;
  uint16_t sw_stop;
} ValueInput;

typedef struct _TimeOnOff {
  uint16_t onTime;
  uint16_t offTime;
  uint16_t targetCount;
  uint16_t lastState;
  uint16_t update;
} TimeOnOff;

typedef struct _ValueOutput {
  TimeOnOff led_green;
  TimeOnOff led_red;
  TimeOnOff buzzer;
  TimeOnOff led_start;
  TimeOnOff led_stop;
  TimeOnOff rel_break;
} ValueOutput;
#pragma pack(pop)

// ASCII, BINARY 타입을 혼용하고 있음
// 하나로 변경할 것
#define SIZE_HEAD   1
#define SIZE_TYPE   1
#define SIZE_TS     4
#define SIZE_DATA_INPUT   5
#define SIZE_DATA_OUTPUT  60
#define SIZE_CRC16  4
#define SIZE_TAIL   1
#define SIZE_TOTAL_INPUT  (SIZE_HEAD+SIZE_TYPE+SIZE_TS+SIZE_DATA_INPUT+SIZE_CRC16+SIZE_TAIL)
#define SIZE_TOTAL_OUTPUT (SIZE_HEAD+SIZE_TYPE+SIZE_TS+SIZE_DATA_OUTPUT+SIZE_CRC16+SIZE_TAIL)
#define IDX_HEAD    0
#define IDX_TYPE          (SIZE_HEAD)
#define IDX_TS            (SIZE_HEAD+SIZE_TYPE)
#define IDX_DATA          (SIZE_HEAD+SIZE_TYPE+SIZE_TS)
#define IDX_CRC16_INPUT   (SIZE_HEAD+SIZE_TYPE+SIZE_TS+SIZE_DATA_INPUT)
#define IDX_CRC16_OUTPUT  (SIZE_HEAD+SIZE_TYPE+SIZE_TS+SIZE_DATA_OUTPUT)
#define IDX_TAIL_INPUT    (SIZE_HEAD+SIZE_TYPE+SIZE_TS+SIZE_DATA_INPUT+SIZE_CRC16)
#define IDX_TAIL_OUTPUT   (SIZE_HEAD+SIZE_TYPE+SIZE_TS+SIZE_DATA_OUTPUT+SIZE_CRC16)
#define DATA_HEAD   0x02
#define DATA_TAIL   0x03

namespace FADUINO {
  enum RELAY {
    OFF, ON
  };
}

enum STATE_ACT {
  INFINITE, ONCE, TWICE, THRICE, T4, T5, T6, T7, T8, T9, T10
};

enum STATE_INPUT {
  RELEASED, PUSHED, DOUBLE, LONG
};

enum STATE_OUTPUT {
  OFF, ON, BLINK_SLOW, BLINK_FAST
};

enum STATE_ESTOP {
  ACTION, IDLE
};

enum FSM_SERIAL {
  HEAD, TYPE, TS, DATA, CRC16, TAIL, OK
};

enum TYPE_CMD {
  UNKNOWN, CMD, ACK, RES, DONE, SENSOR
};

class CRC {
public:
  static unsigned short CRC16(unsigned char data[], unsigned short length) {
    /* CRC upper byte value table */
    static unsigned char CRCHiBytes[] = {
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
    0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40};

    /* CRC lower byte value table */
    static unsigned char CRCLoBytes[] = {
    0x00, 0xC0, 0xC1, 0x01, 0xC3, 0x03, 0x02, 0xC2, 0xC6, 0x06, 0x07, 0xC7, 0x05, 0xC5, 0xC4, 0x04,
    0xCC, 0x0C, 0x0D, 0xCD, 0x0F, 0xCF, 0xCE, 0x0E, 0x0A, 0xCA, 0xCB, 0x0B, 0xC9, 0x09, 0x08, 0xC8,
    0xD8, 0x18, 0x19, 0xD9, 0x1B, 0xDB, 0xDA, 0x1A, 0x1E, 0xDE, 0xDF, 0x1F, 0xDD, 0x1D, 0x1C, 0xDC,
    0x14, 0xD4, 0xD5, 0x15, 0xD7, 0x17, 0x16, 0xD6, 0xD2, 0x12, 0x13, 0xD3, 0x11, 0xD1, 0xD0, 0x10,
    0xF0, 0x30, 0x31, 0xF1, 0x33, 0xF3, 0xF2, 0x32, 0x36, 0xF6, 0xF7, 0x37, 0xF5, 0x35, 0x34, 0xF4,
    0x3C, 0xFC, 0xFD, 0x3D, 0xFF, 0x3F, 0x3E, 0xFE, 0xFA, 0x3A, 0x3B, 0xFB, 0x39, 0xF9, 0xF8, 0x38,
    0x28, 0xE8, 0xE9, 0x29, 0xEB, 0x2B, 0x2A, 0xEA, 0xEE, 0x2E, 0x2F, 0xEF, 0x2D, 0xED, 0xEC, 0x2C,
    0xE4, 0x24, 0x25, 0xE5, 0x27, 0xE7, 0xE6, 0x26, 0x22, 0xE2, 0xE3, 0x23, 0xE1, 0x21, 0x20, 0xE0,
    0xA0, 0x60, 0x61, 0xA1, 0x63, 0xA3, 0xA2, 0x62, 0x66, 0xA6, 0xA7, 0x67, 0xA5, 0x65, 0x64, 0xA4,
    0x6C, 0xAC, 0xAD, 0x6D, 0xAF, 0x6F, 0x6E, 0xAE, 0xAA, 0x6A, 0x6B, 0xAB, 0x69, 0xA9, 0xA8, 0x68,
    0x78, 0xB8, 0xB9, 0x79, 0xBB, 0x7B, 0x7A, 0xBA, 0xBE, 0x7E, 0x7F, 0xBF, 0x7D, 0xBD, 0xBC, 0x7C,
    0xB4, 0x74, 0x75, 0xB5, 0x77, 0xB7, 0xB6, 0x76, 0x72, 0xB2, 0xB3, 0x73, 0xB1, 0x71, 0x70, 0xB0,
    0x50, 0x90, 0x91, 0x51, 0x93, 0x53, 0x52, 0x92, 0x96, 0x56, 0x57, 0x97, 0x55, 0x95, 0x94, 0x54,
    0x9C, 0x5C, 0x5D, 0x9D, 0x5F, 0x9F, 0x9E, 0x5E, 0x5A, 0x9A, 0x9B, 0x5B, 0x99, 0x59, 0x58, 0x98,
    0x88, 0x48, 0x49, 0x89, 0x4B, 0x8B, 0x8A, 0x4A, 0x4E, 0x8E, 0x8F, 0x4F, 0x8D, 0x4D, 0x4C, 0x8C,
    0x44, 0x84, 0x85, 0x45, 0x87, 0x47, 0x46, 0x86, 0x82, 0x42, 0x43, 0x83, 0x41, 0x81, 0x80, 0x40};
    unsigned char HiByte = 0xFF;	// Initialize higher CRC byte
    unsigned char LoByte = 0xFF;	// Initialize lower CRC byte
    unsigned short idx;	// Index in the table

    while (length--) {
        idx = LoByte ^ *data++;
        // Calculate CRC
        LoByte = HiByte ^ CRCHiBytes[idx];
        HiByte = CRCLoBytes[idx];
    }

    return (HiByte << 8 | LoByte);
  }
};